Source/WTF/wtf/text/WYHash.h

```diff
@@ -0,0 +1,275 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#pragma once
+
+#include <wtf/Int128.h>
+#include <wtf/UnalignedAccess.h>
+#include <wtf/text/StringHasher.h>
+
+namespace WTF {
+
+#define FORCE_CONVERT_IN_READ 0
+
+// https://github.com/wangyi-fudan/wyhash
+class WYHash {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static constexpr unsigned flagCount = StringHasher::flagCount;
+    static constexpr unsigned maskHash = StringHasher::maskHash;
+    static constexpr uint64_t secret[4] = { 0xa0761d6478bd642full, 0xe7037ed1a0b428dbull, 0x8ebc6af09c88c6e3ull, 0x589965cc75374cc3ull };
+    typedef StringHasher::DefaultConverter DefaultConverter;
+
+    WYHash() = default;
+
+    template<typename T, typename Converter = DefaultConverter>
+    static constexpr unsigned computeHashAndMaskTop8Bits(const T* data, unsigned characterCount)
+    {
+        return StringHasher::finalizeAndMaskTop8Bits(computeHashImpl<T, Converter>(data, characterCount));
+    }
+
+private:
+    friend class StringHasher;
+
+    ALWAYS_INLINE static constexpr uint64_t wyrot(uint64_t x)
+    {
+        return (x >> 32) | (x << 32);
+    }
+
+    ALWAYS_INLINE static constexpr void wymum(uint64_t* A, uint64_t* B)
+    {
+        UInt128 r = static_cast<UInt128>(*A) * *B;
+        *A = static_cast<uint64_t>(r);
+        *B = static_cast<uint64_t>(r >> 64);
+    }
+
+    ALWAYS_INLINE static constexpr uint64_t wymix(uint64_t A, uint64_t B)
+    {
+        wymum(&A, &B);
+        return A ^ B;
+    }
+
+    template<typename T, typename Converter = DefaultConverter>
+    struct Reader16Bit {
+        ALWAYS_INLINE static constexpr bool hasDefaultConverter()
+        {
+            return std::is_same<Converter, DefaultConverter>::value;
+        }
+
+        ALWAYS_INLINE static constexpr uint32_t convert(const T p)
+        {
+            return Converter::convert(static_cast<uint16_t>(p));
+        }
+
+        ALWAYS_INLINE static constexpr uint64_t wyr3(const T* p)
+        {
+            static_assert(sizeof(T) == 2);
+            uint16_t v = convert(p[0]);
+            uint64_t firstByte = static_cast<uint8_t>(v);
+            uint64_t secondByte = static_cast<uint8_t>(v >> 8);
+            return (firstByte << 16) | (secondByte << 8) | secondByte;
```

Comment:
Does it work for BigEndian?

Reply:
Yes, for `wyr3`, the original implementation uses this for both `BIG_ENDAN` and `LITTLE_ENDAN`.
https://github.com/wangyi-fudan/wyhash/blob/77e50f267fbc7b8e2d09f2d455219adb70ad4749/wyhash.h#L116

Reply:
btw, `BIG_ENDAN` or `LITTLE_ENDAN` only affects `memcpy`.